lib/CPathfinder.cpp

```diff
@@ -79,96 +86,118 @@ void CPathfinder::calculatePaths()
 	//logGlobal->infoStream() << boost::format("Calculating paths for hero %s (adress  %d) of player %d") % hero->name % hero % hero->tempOwner;
 
 	//initial tile - set cost on 0 and add to the queue
-	CGPathNode &initialNode = *getNode(out.hpos);
-	initialNode.turns = 0;
-	initialNode.moveRemains = hero->movement;
-	mq.push_back(&initialNode);
+	CGPathNode * initialNode = out.getNode(out.hpos, hero->boat ? ELayer::SAIL : ELayer::LAND);
+	initialNode->turns = 0;
+	initialNode->moveRemains = hero->movement;
+	pq.push(initialNode);
 
-	while(!mq.empty())
+	while(!pq.empty())
 	{
-		cp = mq.front();
-		mq.pop_front();
+		cp = pq.top();
+		pq.pop();
+		cp->locked = true;
+		ct = &gs->map->getTile(cp->coord);
+		ctObj = ct->topVisitableObj(isSourceInitialPosition());
 
 		int movement = cp->moveRemains, turn = cp->turns;
+		hlp->updateTurnInfo(turn);
 		if(!movement)
 		{
-			movement = maxMovePoints(cp);
-			turn++;
+			hlp->updateTurnInfo(++turn);
+			movement = hlp->getMaxMovePoints(cp->layer);
 		}
 
 		//add accessible neighbouring nodes to the queue
-		addNeighbours(cp->coord);
+		addNeighbours();
 		for(auto & neighbour : neighbours)
 		{
-			dp = getNode(neighbour);
 			dt = &gs->map->getTile(neighbour);
-			useEmbarkCost = 0; //0 - usual movement; 1 - embark; 2 - disembark
+			dtObj = dt->topVisitableObj();
+			for(ELayer i = ELayer::LAND; i <= ELayer::AIR; i.advance(1))
+			{
```

Comment:
Just a notice about performance:
Each CPU cycle inside this (main inner) loop consume 0.5% of overall time.